crates/hyperswitch_domain_models/src/type_encryption.rs

```diff
@@ -0,0 +1,237 @@
+
+        Ok(Self::new(value.into(), encrypted))
+    }
+}
+
+#[async_trait]
+impl<
+        V: crypto::DecodeMessage + crypto::EncodeMessage + Send + 'static,
+        S: masking::Strategy<Vec<u8>> + Send,
+    > TypeEncryption<Vec<u8>, V, S> for crypto::Encryptable<Secret<Vec<u8>, S>>
+{
+    #[instrument(skip_all)]
+    async fn encrypt(
+        masked_data: Secret<Vec<u8>, S>,
+        key: &[u8],
+        crypt_algo: V,
+    ) -> CustomResult<Self, errors::CryptoError> {
+        let encrypted_data = crypt_algo.encode_message(key, masked_data.peek())?;
+
+        Ok(Self::new(masked_data, encrypted_data.into()))
+    }
+
+    #[instrument(skip_all)]
+    async fn decrypt(
+        encrypted_data: Encryption,
+        key: &[u8],
+        crypt_algo: V,
+    ) -> CustomResult<Self, errors::CryptoError> {
+        let encrypted = encrypted_data.into_inner();
+        let data = crypt_algo.decode_message(key, encrypted.clone())?;
+
+        Ok(Self::new(data.into(), encrypted))
+    }
+}
+
+pub trait Lift<U> {
+    type SelfWrapper<T>;
+    type OtherWrapper<T, E>;
+
+    fn lift<Func, E, V>(self, func: Func) -> Self::OtherWrapper<V, E>
+    where
+        Func: Fn(Self::SelfWrapper<U>) -> Self::OtherWrapper<V, E>;
+}
+
+impl<U> Lift<U> for Option<U> {
+    type SelfWrapper<T> = Option<T>;
+    type OtherWrapper<T, E> = CustomResult<Option<T>, E>;
+
+    fn lift<Func, E, V>(self, func: Func) -> Self::OtherWrapper<V, E>
+    where
+        Func: Fn(Self::SelfWrapper<U>) -> Self::OtherWrapper<V, E>,
+    {
+        func(self)
+    }
+}
+
+#[async_trait]
+pub trait AsyncLift<U> {
+    type SelfWrapper<T>;
+    type OtherWrapper<T, E>;
+
+    async fn async_lift<Func, F, E, V>(self, func: Func) -> Self::OtherWrapper<V, E>
+    where
+        Func: Fn(Self::SelfWrapper<U>) -> F + Send + Sync,
+        F: futures::Future<Output = Self::OtherWrapper<V, E>> + Send;
+}
+
+#[async_trait]
+impl<U, V: Lift<U> + Lift<U, SelfWrapper<U> = V> + Send> AsyncLift<U> for V {
+    type SelfWrapper<T> = <V as Lift<U>>::SelfWrapper<T>;
+    type OtherWrapper<T, E> = <V as Lift<U>>::OtherWrapper<T, E>;
+
+    async fn async_lift<Func, F, E, W>(self, func: Func) -> Self::OtherWrapper<W, E>
+    where
+        Func: Fn(Self::SelfWrapper<U>) -> F + Send + Sync,
+        F: futures::Future<Output = Self::OtherWrapper<W, E>> + Send,
+    {
+        func(self).await
+    }
+}
+
+#[inline]
+pub async fn encrypt<E: Clone, S>(
+    inner: Secret<E, S>,
+    key: &[u8],
+) -> CustomResult<crypto::Encryptable<Secret<E, S>>, errors::CryptoError>
+where
+    S: masking::Strategy<E>,
+    crypto::Encryptable<Secret<E, S>>: TypeEncryption<E, crypto::GcmAes256, S>,
+{
+    record_operation_time(
+        crypto::Encryptable::encrypt(inner, key, crypto::GcmAes256),
+        &metrics::ENCRYPTION_TIME,
+        &metrics::CONTEXT,
+        &[],
+    )
+    .await
+}
+
+#[inline]
+pub async fn encrypt_optional<E: Clone, S>(
```

Comment:
We should either name this function differently (say `encrypt_optional_field`?), or have this implemented as a trait on `Secret<E>` and `Option<Secret<E>>`, so that both have the same method name. Having the name as `encrypt_optional` has caused external folks to think that the encryption process itself happens optionally.

Reply:
Since there are some other changes as well in this area taken up by @ArjunKarthik, this can be done in a later PR